### ISSUE TITLE
Should the spec include an interface and bootstrapping for type 'ReactiveStreams'

### ISSUE THREAD

**Issue by hutchig** [eclipse/microprofile-reactive-streams#96]
I would like to discuss:

There is a suggestion above to make the builders, 
(say PublisherBuilder, ProcessorBuilder, SubscriberBuilder) have interfaces.
This is primarliy to enable easier consumption via javadoc and IDEs etc.

As well as different implementations, another thing that interfaces 
enable is different implementations, dynamic-proxies and thus injection, 
interceptors, parts of aop etc.

Is there a valid use case here that interfaces would enable?

Say one looks for a potential place to inject/proxy/intercept etc an
impl of  

PublisherBuilder
ProcessorBuilder
SubscriberBuilder

One gets to where one initially gets hold of them.
So starting with PublisherBuilder, the places are: 

1 - Its two constructors
2 - ReactiveStreams https://github.com/eclipse/microprofile-reactive-streams/blob/master/api/src/main/java/org/eclipse/microprofile/reactive/streams/ReactiveStreams.java

So...
1 - Obviously the constructors in 1 have a concrete type that is not replaceable.
2 - Is a concrete class with a fixed implementation, albeit I imagine a sensible and wise one, that
also returns concrete classes and so on and so on until the 'run' from a non-concrete
loadable/injectable ReactiveStreamsEngine.

At the moment we have an interface and bootstrapping method (service loading and injection mechanism) in the spec for ReactiveStreamsEngine.

I suggest we should explore having an interface and bootstrapping method for each 'way in' e.g. ReactiveStreams.java
to enable independent implementations or a good reason not to.

**Comment by jroper** [comment]
So I've started working on this today, and what I'm really not sure about is whether having this completely negates the need for `ReactiveStreamsEngine` interface, and `Graph` and `Stage` and all the subclasses of that.

If we keep them, that does mean that the code that runs the graph controls not only which engine is used, but all the runtime attributes of it, eg the context and thread pools it runs in. In contrast, if we get rid of them, then the code that builds the graph is in control, and the code that runs the graph is only able to invoke `build` and `run`.

Typically, the code that builds the graph is user code, and the code that runs the graph is framework code. So maybe it's fine to keep the spi because it's better to put the framework in control. But it does feel very odd having a double SPI.

**Comment by hutchig** [comment]
I think this has been addressed under the work in #98 with ReactiveStreamsFactory.java

**Comment by cescoffier** [comment]
Agreed, it's been covered. However, it should be documented in the spec.
